### ISSUE TITLE
Hi system wait times with IO. Many file operations to index* and temp-index-sorter*

### ISSUE THREAD

**Issue by agattis** [syncthing/syncthing#3978]
When connected to  5 peers, with three folders totaling around 24K files/folders, and one system in send-only mode, system IO wait time on all peers gradually increases to 95% with syncthing opening many db files in the config folder. System becomes very slow to respond.

Reproducible when new hosts are added. Said hosts have a copy of files already and scan finishes. Some changes are detected but the send-only host invokes 'Override Changes' though override action seems enable to finish as status never changes (i.e. the button never goes away).

All hosts are Ubuntu 16.04 with syncthing installed via the repo as on the author's site instructions. Hardware is i5-3470 and 4G RAM at min. No other processes of significance running

Persists between reboots and the most recent upgrade from 0.14.22 to 0.14.23 (current swarm is mixed).

Interestingly, with two hosts I did not see this behavior. Possibility it occurred as new peers were added at the newer version before the old peers updated, but all are affected currently

Sample of IOTOP - many pages

syncthing 1060 2797            allen   11r      REG                8,1       3613811   24517752 /home/allen/.config/syncthing/index-v0.14.0.db/044217.ldb
syncthing 1060 2797            allen   12w      REG                8,1        163166   25431088 /home/allen/.config/syncthing/tmp-index-sorter.306530046/MANIFEST-000000
syncthing 1060 2797            allen   13w      REG                8,1       2084053   26088947 /home/allen/.config/syncthing/tmp-index-sorter.448111433/003109.ldb
syncthing 1060 2797            allen   14r      REG                8,1       2109389   24514888 /home/allen/.config/syncthing/index-v0.14.0.db/343072.ldb
syncthing 1060 2797            allen   15u     sock                0,8           0t0      20839 protocol: TCP
syncthing 1060 2797            allen   16uW     REG                8,1             0   25826702 /home/allen/.config/syncthing/tmp-index-sorter.615775423/LOCK
syncthing 1060 2797            allen   17r      REG                8,1       2343886   26347355 /home/allen/.config/syncthing/tmp-index-sorter.097130083/002484.ldb
syncthing 1060 2797            allen   18w      REG                8,1        135558   25300232 /home/allen/.config/syncthing/tmp-index-sorter.217008987/MANIFEST-000000
syncthing 1060 2797            allen   19w      REG                8,1       1386043   25037809 /home/allen/.config/syncthing/tmp-index-sorter.586510721/006650.ldb
syncthing 1060 2797            allen   20w      REG                8,1        583181   25171151 /home/allen/.config/syncthing/tmp-index-sorter.820494060/006577.ldb
syncthing 1060 2797            allen   21w      REG                8,1             0   25300252 /home/allen/.config/syncthing/tmp-index-sorter.217008987/002975.log
syncthing 1060 2797            allen   22r      REG                8,1       2185625   24648341 /home/allen/.config/syncthing/tmp-index-sorter.186965654/006725.ldb
syncthing 1060 2797            allen   23r      REG                8,1       2140281   24515232 /home/allen/.config/syncthing/index-v0.14.0.db/336000.ldb
syncthing 1060 2797            allen   24w      REG                8,1       1543296   26480161 /home/allen/.config/syncthing/tmp-index-sorter.730459277/000671.ldb
syncthing 1060 2797            allen   25uW     REG                8,1             0   25168797 /home/allen/.config/syncthing/tmp-index-sorter.820494060/LOCK
syncthing 1060 2797            allen   26r      REG                8,1       2176073   24647912 /home/allen/.config/syncthing/tmp-index-sorter.186965654/006721.ldb
syncthing 1060 2797            allen   27r      REG                8,1       2329366   24517748 /home/allen/.config/syncthing/index-v0.14.0.db/044210.ldb
syncthing 1060 2797            allen   28r      REG                8,1        419474   26480239 /home/allen/.config/syncthing/tmp-index-sorter.054421606/000771.ldb
syncthing 1060 2797            allen   29r      REG                8,1       2153746   24522185 /home/allen/.config/syncthing/index-v0.14.0.db/263434.ldb
syncthing 1060 2797            allen   30u     sock                0,8           0t0      21816 protocol: TCP
syncthing 1060 2797            allen   31r      REG                8,1       2140746   24523079 /home/allen/.config/syncthing/index-v0.14.0.db/341887.ldb
syncthing 1060 2797            allen   32r      REG                8,1       2148884   26225024 /home/allen/.config/syncthing/tmp-index-sorter.820058644/003199.ldb
syncthing 1060 2797            allen   33w      REG                8,1       1175893   26223926 /home/allen/.config/syncthing/tmp-index-sorter.820058644/003258.ldb
syncthing 1060 2797            allen   34w      REG                8,1        154880   26479988 /home/allen/.config/syncthing/tmp-index-sorter.054421606/LOG
syncthing 1060 2797            allen   35r      REG                8,1       2059260   24519713 /home/allen/.config/syncthing/index-v0.14.0.db/331024.ldb
syncthing 1060 2797            allen   36uW     REG                8,1             0   26088911 /home/allen/.config/syncthing/tmp-index-sorter.448111433/LOCK

**Comment by calmh** [comment]
This is expected on first connect. A full index exchange is required, and the index needs to be sorted chronologically. For an index of a non-trivial size this means creating a temporary copy on disk.

**Comment by agattis** [comment]
This is after a week of contact. What length of time is expected?

**Comment by calmh** [comment]
You said "Reproducible when new hosts are added."

**Comment by agattis** [comment]
Yes, when adding a new host, it exhibits the same behavior as the other members. Prehapse one should allow one host to finish being 'added' before adding others?

**Comment by calmh** [comment]
Possibly. Are these very slow machines, I/O wise? I have a significantly larger number of files and this stage takes a couple of minutes on my laptop when first connecting to a new device...

**Comment by agattis** [comment]
No, i5 desktops with SATA drives. Reasonable IO

**Comment by calmh** [comment]
Slightly surprising then that it is that slow, but perhaps doing multiple of these in parallel such as it's doing (five index sorting operations per device, if there are six devices in total) makes it spend too much time in disk seeks or so.

You said the devices already have the data on them, which might also make it worse. Essentially one device will send all it's index data, the other side will send all it's index data, they will both need to reconcile the differences and agree on a common version, and then send everything again.

This seems to be something that will just take a while in your setup. Doing it with less machines in parallel is probable better.

**Comment by agattis** [comment]
At first addition, I'm connecting an already rsync'ed copy of the data, and it takes a day or two to build the block hashes, as it's around 6TB. Though initially, that only affected the recently added host.

**Comment by calmh** [comment]
Yeah, it's a large index then. Let things settle, I guess.

**Comment by agattis** [comment]
I'll take the systems down, and re-add one at a time and report back then. That will take week or so.

**Comment by bigbigbang** [comment]
My configuration is a bit the same, and I experience long-time before it start to sync too. My conf is :
1 master node (sent only mode)
7 other nodes. 
First of all, if I have to add a node with already existing data (and no syncthing index) I sync it alone with the 'master' node (and Pause the other nodes). The first sync is quiet long to start as syncthing have to build the block hashes, but it didn't take more than 4 Hours for 2TO. So for 6TO like you, it will be 12h not a day ! (and it is for node with only one core runing at 2.4 Ghz, with 1 Go RAM, DD on sata). With bigger config (16 core, 64 GO RAM), I build the 2To block hashes in less than 2H max.
Then if I add a new (empty) device in the cluster, I experienced a ~30min delay before they all start to send to other device. This said, when the device is empty, their is no building of block hashes.

**Comment by agattis** [comment]
It seems to be tied to pre-scanning. 

Specifically, I started with seven nodes, all identical. 1.6 TB in 1800 files. Node one configured to send-only. Node two, added and scanned itself as expected with IO and CPU as expected. Node three, same. Exiting member nodes show nominal CPU and IO during the addition of the new nodes.

Node four I configured the shared folder with the same ID in advance. Upon introducing to the swarm, all nodes spiked CPU and IO wait to high levels - Both in the 60 range average according to top 

One node showed significantly higher CPU (200%) , but little IO Wait. Turns out that node has an SSD and appears to be performing db transactions more quickly. 

After 30 min or so, the send-only node says 'Out of sync' I hit the override changes button and the nodes spike usage. Additional files files listed as out of sync. Repeat over the next 30 min.

I'll remove the last node and re-add normally to see what the affect is.

**Comment by bigbigbang** [comment]
Did you sync nodes that have already data existing alone with the master ? (the node on sent-only). You must pause all other nodes, if you want to sync a node with already existing data. Otherwise, in your configuration, when you start, the node 4 , node 2 and 3 are downloading the 'extra' data from node 4.

**Comment by agattis** [comment]
Indeed - all hosts have preexisting identical copies. I expected new additions to scan themselves, establish local hashes, compare that to the global state and recognize no action was needed. What happens is something else that triggers a lot of DB activity on all swarm members.

I did as the previous commenter suggested and upon adding a new node, I paused it's connections to the other members of the swarm (from within the new node's Remote Devices section), leaving only it's connection to the Read Only member.

Activity dropped across all other members to nominal levels. Even on the RO node.

Is this the expected behavior upon adding new nodes with existing copies of data ?

**Comment by bigbigbang** [comment]
So is it working like expected now ?
When your node with existing data is sync with the node on sent-only mode, you can re-start (resume) other nodes. They will exchange a bit of data (to be sure they are in sync) but not trigger high db activity.

> Activity dropped across all other members to nominal levels. Even on the RO node.
Yes 'slaves' nodes are paused, and your master only sent data to 1 node. Your 'slave' node here is calculating hash of data-block, and compare the result with what the master have. As they have the same data block , they is no intense activity.

I don't know if this is a good (and even valid) answer , i'm still a newbie too. But as I was in the same situation, this is how I have resolved it.

**Comment by calmh** [comment]
It's not so much about existing copies of data, but about having a large index to transmit the first time they connect (and subsequent times until completed). The operation is quite heavy unfortunately as the index needs to be sorted in a temporary database.

I can imagine that when this happens in both directions at once there is increased load as each side is copying and sorting the index at the same time as it's receiving a copy of it from the other side and committing that to the same database in parallel.

It becomes worse when the devices have the same data but independently scanned. The reason is that we track version vectors for each file, to be able to follow it's evolution over time and resolve conflicts. Your two devices have all their files in conflict when they first meet, as they have the same files but there is no shared history between the two copies. It's like if you wrote a document and your colleague wrote a document with the same name, and now you need to merge them. So for every file in the index, the following happens:

- Device A to device B: I have a file `f`, with version `A:1`
- Device B to device A: I have a file `f`, with version `B:1`
- A to B: I have resolved the conflict for file `f`, here is the new version `A:1, B:1`
- B to A: I have successfully handled your conflict resolution, my version is now `A:1, B:1`

So that's four roundtrips per file.

**Comment by agattis** [comment]
Yes, this seems to be the case. Upon adding a new host with existing data, as long as I pause the connections to the other members things proceed as expected.  I can add multiple hosts as long as they both only communicate with a single member who is configured to send-only. At the end, I can 'Override Changes' to get the file versions accepted, and unpause the other members from the newly added hosts.

This use case - pre-seeding - might be a common one based on my own experience of transporting many terabytes to remote sites with poor throughput in rural areas. Mailing a hard drive is much cheaper than a whole system. 

Is a more graceful approach to pre-seeding, perhaps worthy of a feature request?

**Comment by agattis** [comment]
I made a mistake when adding one of the successive nodes and didn't pause one of the hosts. It may have been added a host when I was looking as the one I left it connected to was an introducer, but regardless, it send all the other nodes into an index-update mode and brought other processes to their knees. I backed the new node out, but we're still doing index work after a day and I may have to just take them down again and repeat.

Since this is expected (just surprising) behavior, I'll call this closed and start reading the forums.

Thanks.
